Allow more control over which specs are reused

alalazo opened the pull request:
fixes #42813
fixes #32888

This PR gives users finer control over which specs are reused during concretization.

The value of the `concretizer:reuse` config option now can take an object with the following properties:
- `roots`: true if reusing roots, false if reusing just dependencies
- `exclude`: list of constraints used to select reusable specs 
- `include`: list of constraints used to select reusable specs 
- `from`: allows to select the sources of reused specs

### Examples

#### Reuse only specs compiled with GCC
```yaml
concretizer:
  reuse:
    roots: true
    include:
    - "%gcc"
```

#### `openmpi` must be used from externals, and it must be the only external used
```yaml
concretizer:
  reuse:
    roots: true
    from:
    - type: local
      exclude:
      - "openmpi"
    - type: buildcache
      exclude:
      - "openmpi"
    - type: external
      include:
      - "openmpi"
```


### Extensions in future PRs

Since the items in the `from:` list are object, they can be extended easily. As an example we might allow something like:
```yaml
concretizer:
  reuse:
    from:
    - type: buildcache
      name: e4s
```
to select specs only from a specific buildcache. Or select specs based on installation dates etc.

alalazo committed:
ASP-based solver: allow more control over specs that are reused

alalazo committed:
Allow filtering specs to be reused

alalazo committed:
Allow selecting sources for reused specs

alalazo committed:
Add documentation

alalazo committed:
Add type hints to the new class

alalazo committed:
Do not test expected lengths on Windows

Windows, for some reason, has a different test database

alalazo committed:
Add exclude attribute, allow default values in configuration

alalazo committed:
Add support for external spec selection

alalazo committed:
Fix include/exclude semantic, add a unit-test

alalazo committed:
Rename "reuse:strategy" to "reuse:roots"

alalazo commented:
@haampie Changed `reuse:strategy: to `reuse:roots` as we discussed in DMs

haampie commented:
Can we still change the name of `mirror` to `buildcache`? In our docs we say

> Build caches refer to mirrors that provide prebuilt packages.

so I think that's more accurate.

psakievich commented on lib/spack/spack/schema/concretizer.py:
```diff
@@ -9,13 +9,40 @@
+                                    "properties": {
+                                        "type": {
+                                            "type": "string",
+                                            "enum": ["local", "mirror", "external"],
```
I know I'm late to the review, but how much of a lift would it be to add `upstreams` to this list? Is the intent that those would be coupled under `local`?  I would personally like to see them differentiated if possible.

alalazo replied:
> I would personally like to see them differentiated if possible.

We can have additional PRs later, to add more control - but I'll keep this one as is[^1] :slightly_smiling_face: What would be the use case for distinguishing upstreams from local? 

[^1]: There are also other extensions mentioned in the description that I'd like to add later, for instance selecting a mirror by name, when multiple mirrors are registered.

psakievich replied:
Primarily to have more control. Say I suspect there might be an issue with the upstream, or I wanted to test phasing one out. Follow on is fine by me.

tgamblin replied:
See #42999 -- a symmetric config option here makes a lot of sense. Upstreams have names and paths that they can be identified by (kind of like mirrors)

alalazo committed:
Fix issues after rebase

Our mock repository doesn't have mock runtime
logic, so we need to monkeypatch the
'_has_runtime_dependencies' function

alalazo committed:
Rename "mirror" -> "buildcache"

alalazo commented:
> Can we still change the name of type: mirror to type: buildcache? In our docs we say

See https://github.com/spack/spack/pull/42782/commits/fbc5edb13858aea0e15bcd91f80deef6b1acd308

alalazo committed:
Update after rebase

tgamblin commented on lib/spack/spack/schema/concretizer.py:
```diff
@@ -9,13 +9,40 @@
+                            "exclude": LIST_OF_SPECS,
+                            "from": {
+                                "type": "array",
+                                "items": {
+                                    "type": "object",
+                                    "properties": {
+                                        "type": {
+                                            "type": "string",
+                                            "enum": ["local", "buildcache", "external"],
+                                        },
+                                        "include": LIST_OF_SPECS,
+                                        "exclude": LIST_OF_SPECS,
+                                    },
+                                },
```
would really like to be able to write:

```yaml
concretizer:
    reuse:
        from: ["local", "buildcache", "external"]
```

is there an issue w/adding this?

The other thing that is missing from this vs. our idea for `from:` in `packages.yaml` is that there, we were going to specifically allow someone to say `from: "source"`.  How would you say that you *always* want some package to be built from source and never reused here?

psakievich replied:
> How would you say that you always want some package to be built from source and never reused here?

Perhaps we could add a global exclusion?
``` yaml
concretizer:
  reuse:
    roots: true
    from:
    - type: local
      exclude:
      - "openmpi"
    - type: buildcache
      exclude:
      - "openmpi"
    - type: external
      include:
      - "openmpi"
    always_exclude:
    - "end_application_a"
    - "end_application_b"
 ```
 or
``` yaml
concretizer:
  reuse:
    roots: true
    from:
    - type: local
      exclude:
      - "openmpi"
    - type: buildcache
      exclude:
      - "openmpi"
    - type: external
      include:
      - "openmpi"
  never_reuse:
    - "end_application_a"
    - "end_application_b"
 ```
 I don't really like my suggested names, but this interface would work for me.

alalazo replied:
> How would you say that you always want some package to be built from source and never reused here?

@psakievich @tgamblin There is global exclusion, and imo it's more general than the suggested options:
```yaml
concretizer:
  reuse:
    exclude:
    - "foo"
```
With the above `foo` is not reused from local, buildcache or externals. The only way not to build is to get an accidental hash match. Note that we can also do things like:
```yaml
concretizer:
  reuse:
    exclude:
    - "%oneapi"
```
e.g. rebuild all the specs compiled with `oneapi` etc.

psakievich replied:
🤦‍♂️ well that's what I get for commenting without re-checking the code.

tgamblin replied:
ok and `reuse: false` works for everything from source and/or hash?

tgamblin replied:
@alalazo: if you do this does it always build from source?  One thing that people had said they wanted was the ability to *always* build a package from source (even if it's installed).  I would not hold this PR up for it but just curious.

`reuse: false` still forces `--fresh` all the time but will do hash reuse, yes?

alalazo replied:
> reuse: false still forces --fresh all the time but will do hash reuse, yes?

Yes. Also, in the case above we'll always do hash reuse (i.e. if the final hash is installed already, we won't override it).

An option to always rebuild a spec seems more related to installing, than to reusing specs in concretization :thinking:

psakievich replied:
@tgamblin would it need to be **always**?  Like if it is an exact hash match why rebuild?  That would lead to install prefix collisions and would need overwriting etc as I believe @alalazo is alluding to.

 The only exception I can think of is for branch based versions, but that should go away when we add the git-sha pairing to branch versions at concretization. I have a minimum viable product working for that in `spack-manager` that I will need to hook up to the concretizer after we merge #43859.

tgamblin replied:
@psakievich People wanted that, but I don't see why we need to hold up the PR over it.
